### ISSUE TITLE
Direct entry of talk group number

### ISSUE THREAD

**Issue by rogerclarkmelbourne** [rogerclarkmelbourne/Radioddity_GD-77#23]
The MD380Toolz has a feature to allow the direct entry of a Talkgroup number via the keypad.

This is a very useful feature and we should try to implement it in the GD-77

On the MD380 they seem to have modified the Private Call number entry system to handle this, because to enter a TG number you press hash hash before entering the number , and the normal operation is to just press hash once to enter a Private Call number.

I think the easiest way to initially add this feature is to do something similar except, perhaps use the star key as an enter key, ( or another key)

We already know where the current channel data is stored in memory and I successfully changed the Tx TG in the radio by using the debugger to change the value in memory.

However, I checked what happens to that memory location when a private call number is entered and the firmware does not update that memory address until the PTT is pressed.
Which is not that useful in this context.

Looking at the keypad handling, there appear to be tests ( if statements) , which test for keycodes that the GD-77 can’t generate, so if this is the case, it may be possible to repurpose one of the if statements and change the key code it checks for and repurpose the function it calls etc.

The keypad handling only seems to be in the main task, so I presume that the firmware must have a variable , which holds the current UI state, and the key scan code for each key must check that variable and decide what to do under various conditions ??

**Comment by rogerclarkmelbourne** [comment]
This function 0002cea8 seem so check the character and call sprintf

It only seems to be called when the private call number is being entered

**Comment by rogerclarkmelbourne** [comment]
When numbers are entered their ascii codes are stored in DAT_1fff356c

**Comment by forkoz** [comment]
I think this is input length: input_length_1FFFF946

**Comment by rogerclarkmelbourne** [comment]
FUN_0002ce82 seems to do some processing of the number when the PTT is pressed

**Comment by rogerclarkmelbourne** [comment]
In fact 2c008 only seems to be called when in Private call mode and the PTT is pressed

**Comment by rogerclarkmelbourne** [comment]
One way to add this functionality is to possibly handle an extra keypad key, by changing

```
                             LAB_0002d02c                                    XREF[8]:     0002ceea(j), 0002cf18(j), 
                                                                                          0002cf34(j), 0002cf4a(j), 
                                                                                          0002cf52(j), 0002cf6a(j), 
                                                                                          0002cf6e(j), 0002d022(j)  
        0002d02c ff f7 00 ff     bl         FUN_0002ce30                                     undefined FUN_0002ce30()
```

To call a patched function, which then calls checks what the key is, and either handles the key e.g. teh # key, and then calls FUN_0002ce30

However an alternative is to patch this code instead



```
                             LAB_00023486                                    XREF[1]:     00023480(j)  
        00023486 18 f0 57 fb     bl         FUN_0003bb38                                     undefined FUN_0003bb38()
        0002348a 00 28           cmp        r0,#0x0
        0002348c 08 d1           bne        LAB_000234a0
        0002348e 19 f0 31 fa     bl         FUN_0003c8f4                                     undefined FUN_0003c8f4()
        00023492 00 28           cmp        r0,#0x0
        00023494 04 d1           bne        LAB_000234a0
        00023496 ff f7 99 ff     bl         FUN_000233cc                                     undefined FUN_000233cc()
        0002349a bd e8 31 40     pop.w      {r0, r4, r5, lr }
        0002349e 1c e7           b          FUN_000232da                                     undefined FUN_000232da()
                             -- Flow Override: CALL_RETURN (CALL_TERMINATOR)
```
As its a bit higher up in the key scanning heirachy, and appears to be the code which handles the right menu press, when the radio is in PC number entry mode.

i.e This code seems to take the radio out of PC mode and back into normal mode

**Comment by rogerclarkmelbourne** [comment]
umm

I'm not sure how the code at 0x23486 works

Especially 2349a, as its poping lr at the end not PC

The other exit from this function is 

 ```
                            LAB_00023578                                    XREF[5]:     00023552(j), 00023556(j), 
                                                                                          0002355a(j), 0002355e(j), 
                                                                                          00023562(j)  
        00023578 04 20           mov        r0,#0x4
        0002357a 84 f8 2d 00     strb.w     r0,[r4,#0x2d]=>DAT_1fff5611                      = 04h
                             LAB_0002357e                                    XREF[3]:     000234b0(j), 000234d0(j), 
                                                                                          00023540(j)  
        0002357e 31 bd           pop        { r0, r4, r5, pc }
```

Which pops PC

What happens when you pop lr not PC ?


I thought that the code at 23486 was what terminated the Private Call user interface,

but putting the same code into my new patch function

 ```
                            **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             undefined4 __stdcall Patch_keypad_input_handler_000653B8
             undefined4        r0:4           <RETURN>
             undefined4        Stack[-0xc]:4  local_c
             undefined4        Stack[-0x10]:4 local_10
             undefined4        Stack[-0x14]:4 local_14
             undefined4        Stack[-0x18]:4 local_18
             undefined4        Stack[-0x28]:4 local_28
                             Patch_keypad_input_handler_000653B8+1           XREF[1,2]:   FUN_0002342a:0002342c(c), 
                             Patch_keypad_input_handler_000653B8                          FUN_00065518:00065526(*), 
                                                                                          00065548(*)  
        000653b8 02 b5           push       { r1, lr }
        000653ba 00 bf           nop
        000653bc 10 4d           ldr        r5,[PTR_DAT_00065400]                            = 1fff197c
        000653be 68 80           strh       r0,[r5,#0x2]=>DAT_1fff197e                       = 0120h
        000653c0 10 4c           ldr        r4,[PTR_DAT_00065404]                            = 1fff55e4
        000653c2 00 bf           nop
        000653c4 4f f4 85 71     mov.w      r1,#0x10a
        000653c8 08 45           cmp        r0,r1
        000653ca 17 d1           bne        LAB_000653fc
        000653cc e0 78           ldrb       r0,[r4,#0x3]=>DAT_1fff55e7                       = 9Fh
        000653ce 01 00           lsl        r1,r0,#0x0
        000653d0 14 d0           beq        LAB_000653fc
        000653d2 00 bf           nop
        000653d4 00 bf           nop
        000653d6 00 bf           nop
        000653d8 00 bf           nop
        000653da 00 bf           nop
        000653dc 00 bf           nop
        000653de 00 bf           nop
        000653e0 00 bf           nop
        000653e2 00 bf           nop
        000653e4 00 bf           nop
        000653e6 00 bf           nop
        000653e8 00 bf           nop
        000653ea d6 f7 a5 fb     bl         FUN_0003bb38                                     undefined FUN_0003bb38()
        000653ee 00 28           cmp        r0,#0x0
        000653f0 04 d1           bne        LAB_000653fc
        000653f2 d7 f7 7f fa     bl         FUN_0003c8f4                                     undefined FUN_0003c8f4()
        000653f6 01 d1           bne        LAB_000653fc
        000653f8 bd f7 e8 ff     bl         FUN_000233cc                                     undefined FUN_000233cc()
                             LAB_000653fc                                    XREF[4]:     000653ca(j), 000653d0(j), 
                                                                                          000653f0(j), 000653f6(j)  
        000653fc 02 bd           pop        { r1, pc }
        000653fe 00 bf           nop
                             PTR_DAT_00065400                                XREF[1]:     000653bc(R)  
        00065400 7c 19 ff 1f     addr       DAT_1fff197c
                             PTR_DAT_00065404                                XREF[1]:     000653c0(R)  
        00065404 e4 55 ff 1f     addr       DAT_1fff55e4
        00065408 00              ??         00h

```

and calling the patch function at the top of ..

```

                             **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             undefined FUN_0002342a()
             undefined         r0:1           <RETURN>
                             FUN_0002342a
        0002342a 38 b5           push       { r3, r4, r5, lr }
        0002342c 41 f0 c4 ff     bl         Patch_keypad_input_handler_000653B8              undefined4 Patch_keypad_input_ha.....

```

Didnt work

My patch function partially works..

It only gets to

` 000653d2 00 bf           nop`

If you are in the PC number entry UI, and then press #14 

But what I then need to do is

1. Read the number and put into the TX TG (this probably won't be too hard, because this is what happens when the PC is pressed when the radio is in Private Call entry mode, except a flag to say its in PC must be set (possibly in the upper byte of the Tx TG 32 bit value.

But I also need to exit from the Private Call UI, which is what I was trying to achieve with this.

Its also possible that the code at 0x23486 doesn't exit from Private Call UI, but I suspect its at least part of the process

Maybe the secret is 

```
        0002349a bd e8 31 40     pop.w      {r0, r4, r5, lr }
        0002349e 1c e7           b          FUN_000232da     
```

and the pop.w is not causing the function to exit at all ?

I'll need to put a breakpoint on it tomorrow

**Comment by rogerclarkmelbourne** [comment]
I just had a thought about how I may be able to exist from the Private Call UI.

If in my patch function, I put the key code of the right arrow key press into r0 , and also into the address it seems to store the last key press, the calling function will act as if the user has pressed the right arrow to exit PC mode.

I will also investigate how the red manu key can exit from Private call mode, as this may not be handled in these functions at all.

But, I do not want the code to restore the Tx Tg from the codeplug value, I want it to retain the value I will put into its memory address

**Comment by rogerclarkmelbourne** [comment]
FYI.

I'm making some progress with this.

What's currently working is that it can detect when the hash key is pressed when in Private Call entry mode.
As pressing hash will be the 'enter' key, I then need to exit from PC entry mode, and I found the best way to do this seems to be to fake the Right Arrow button being pressed, when the user presses the hash

I found that there is a global buf 0x1fff356c which seems to hold the numbers that have been entered but its in ASCII.

I dumped the memory, but I can't see anywhere that the number that has been entered is stored as the number as a 32 bit int, or even as for example "12345" as 0x01,0x02,0x03,0x04,0x05

There must be some existing code that decodes the "12345" string to an int, but at the moment I can't find it. I even looked in the full C decompile you did, and can't find the sort of thing I would expect to see e.g.
```
int total =0;
while(currentChar!=0)
{
total = (total  *10 )+ *currentChar- '0';
currentChar++;
}
```

**Comment by rogerclarkmelbourne** [comment]
Some progress with this, but also some problems :-(

[Patch_keypad_input_handler_000653B8.zip](https://github.com/rogerclarkmelbourne/Radioddity_GD-77/files/3058311/Patch_keypad_input_handler_000653B8.zip)

This gets called from

```
                             **************************************************************
                             *                          FUNCTION                          *
                             **************************************************************
                             undefined FUN_0002342a()
             undefined         r0:1           <RETURN>
                             FUN_0002342a
        0002342a 38 b5           push       { r3, r4, r5, lr }
        0002342c 41 f0 c4 ff     bl         Patch_keypad_input_handler_000653B8              undefined4 Patch_keypad_input_ha
        00023430 00 bf           nop
        00023432 00 bf           nop
```

This does its job...

If in Private Call entry mode, you enter a number and then press the hash key, it parses the number thats been entered and changes the active TG memory location, and also changes the Channel display.

However, although when you first pull the PTT, it does transmit on the talk group number you entered.

A few seconds after you release the PTT, some other part of the code changes the TG back again to the value from the value from the codeplug.

Also although the name that is displayed when not transmitting does not change, when you press the PTT, because the TG has been set back to its codeplug value, when you pull the again, it displays the channel name and from the codeplug, and transmits on the codeplug TG.

The strange thing is that the firmware seems to lookup the name of the channel from the value in the TalkGroup memory.

Also, it doesn't matter if that TG is in the current zone or not !

I think what I may need to do is to create a new global variable, which I set when in the direct TG mode is enabled.

Also. I've just noticed that I should checked that there is a non value number, been entered before applying it to the TG, but thats the least of the problems ;-)

**Comment by rogerclarkmelbourne** [comment]
Thinking about it. Didnt the CPU ID Key checking code use a global to store something.

I'll need to go back to an old version of the code, which still has those functions in it, and see what memory they use.

**Comment by rogerclarkmelbourne** [comment]
It looks like this is the code which causes the TG to get changed back

```
        00018fae 0e 20           mov        r0,#0xe
        00018fb0 02 f0 f8 fd     bl         FUN_0001bba4                                     undefined FUN_0001bba4()

```

So if I can find a variable to flag that the radio is in Direct TG mode, I can patch this call to FUN_0001bba4  to do the call via another small function which will check the flag

However, I'll also need to work out when to clear the flag e.g. when the user changes channel or zone

**Comment by forkoz** [comment]
That function gets the CPUID but it doesn't compare it to anything, it was used by CPUID_read_delay_4CCDE which uses it for some sort of delay. At least we finally get to name some stuff.

**Comment by rogerclarkmelbourne** [comment]
I suppose I could put a value into that global variable address and see if I can notice any changes to the operation of the radio

Or perhaps NOP out any code which reads it.

I did try setting a Read Watch on that address, but GDB kept on reporting some sort of hard fault errors on the CPU.
So that was a dead end.

**Comment by rogerclarkmelbourne** [comment]
BTW. The firmware behaves strangely with reference up updating the memory variable which contains the current channel name string.

I change this to read TGxxxx, but strangely if I change channel and come back to the modified channel, I think it seems to retain this as the new name for the channel.

I will need to double check what’s going on...

I can paste a complete SGL file if you want, but like O said, the effect of the TG number change, only stays until a few seconds after the PTT is released :-(

**Comment by forkoz** [comment]
Wow, they made this harder than it should be. It may be saving names and stuff to the eeprom too which makes things worse. Maybe its worth stepping through the private call function to see how it behaves?

**Comment by rogerclarkmelbourne** [comment]
I already looked at some of the PC code.

Strangely, it only updates the PC ID and the flag byte when you push the PTT, and reverts the TG and the flag when you release the PTT.

Potentially, the problem I'm seeing is an artifact of the post PTT,  PC handling kicking in, even though the radio is not in PC mode.


There seems to be a Mode variable at DAT_1fff197e 

e.g. which gets checked in this code in the keypad handler, in multiple places

e.g this is the existing code that handles the Right Arrow button checking, where #0x10f is the code for the right arrow button, and I think the compiler is using the lsl instruction for some unknown reason, to check if that variable is zero 

```
                             LAB_0002346e                                    XREF[1]:     00023438(j)  
        0002346e 68 88           ldrh       r0,[r5,#0x2]=>DAT_1fff197e                       = 0120h
        00023470 40 f2 0f 11     movw       r1,#0x10f
        00023474 88 42           cmp        r0,r1
        00023476 13 d1           bne        LAB_000234a0
        00023478 e0 78           ldrb       r0,[r4,#0x3]=>DAT_1fff55e7                       = 9Fh
        0002347a 01 00           lsl        r1,r0,#0x0
        0002347c 01 d0           beq        LAB_00023482
        0002347e 98 29           cmp        r1,#0x98
        00023480 01 db           blt        LAB_00023486
```

I do the same check in my code to see if its in private call data entry mode .

Actually, I've just spotted a bug in my code, the check is only for non zero, and it looks like then the menu's are being displayed that value is non zero.

I think value of 0x0E means private call mode, but I'll double check and update my code.


But... Yes. they did make this a lot harder to implement than I thought it would be, mainly because of the sloppy way the existing code works.

**Comment by rogerclarkmelbourne** [comment]
Yep.

Looks like code 0x0e = PC entry

I fixed that small bug in my code

```
        000653ca 0e 28           cmp        r0,#0xe
        000653cc 26 d1           bne        LAB_0006541c
```

**Comment by forkoz** [comment]
What are the other modes?

```
 if ( HIBYTE(Call_mode_1FFF55E6) != 0x3E// private call
            && HIBYTE(Call_mode_1FFF55E6) != 0x3C
            && HIBYTE(Call_mode_1FFF55E6) != 0x43
            && HIBYTE(Call_mode_1FFF55E6) != 0x45
            && HIBYTE(Call_mode_1FFF55E6) != 0x47 )
```

**Comment by rogerclarkmelbourne** [comment]
Interesting

1fff55e7  
0x0e seems to be private call


0x01 = top level of the menus e.g "Contact", "Message" etc being displayed
0x02 = second level of menus
0x0c = Contact menu (listing the contacts)
0x0d = Contact menu - Entering a number for a contact (probably the PC /TG number)
0x0e = Contact menu "Manual dial"

0x03 = Message menu
0x0f = Message menu - inbox


ah.

So when I see 0x0e via the # key, what its actually doing is shortcut to the manual dial menu

It would take me ages to find all the codes.

**Comment by rogerclarkmelbourne** [comment]
I had a look at those functions, but none of them have anything that multiples the value by 10
integer multiply on this MCU will be a single instruction

But I guess the compiler could be presuming that it doesn't have fast integer multiplication and could be doing a  x << 3 + x << 1

**Comment by forkoz** [comment]
So what ends up in 1FFF3590? Can't find anything that turns the characters into integers, all just storts and search and storing them into other places.

**Comment by rogerclarkmelbourne** [comment]
possibly a temporary buffer

Both the memory dumps took recently had zeros in that area

**Comment by forkoz** [comment]
Does it let you view the memory live?

**Comment by rogerclarkmelbourne** [comment]
Yes. I can view live memory

e.g. 

print {char[16]} *0x1fff3590

gives me a print as a string of that memory

When in PC mode its the number I've entered

**Comment by rogerclarkmelbourne** [comment]
I'm using the copy at 1FFF356c
I didn't notice the other copy at 1FFF3590

**Comment by rogerclarkmelbourne** [comment]
The code that changes the TG is in this switch statement

```
                             switchD_0001bbb4::caseD_e                       XREF[1]:     0001bbb4(j)  
        0001be16 48 46           mov        r0,r9
```

Its called immediately when the PTT is pressed when in Private Call mode
Its  **always** all called again a few seconds after the PTT is released, no matter whether the radio is in private call mode or not.

There must be some sort of timer which is started when the PTT is released, because this code doesn't get called straight away.

**Comment by forkoz** [comment]
It runs the parse search functions on that copy a bunch. But I don't find anything that turns it into ints. Unless its done by one of the functions in sub_2D034

**Comment by forkoz** [comment]
the sub? sub_1B228

It does a lot of mulitplying:

```
unsigned int __fastcall sub_1B228(int a1)
{
  unsigned int v1; // r1@1
  char v2; // r2@1
  char v3; // t1@1

  v1 = *(_BYTE *)(a1 + 3);
  v2 = *(_BYTE *)(a1 + 2);
  v3 = *(_BYTE *)a1;
  _R3 = *(_BYTE *)(a1 + 1) & 0xF;
  _R4 = 10000;
  __asm { SMLABB.W R3, R4, R3, R0 ; Signed multiply-accumulate (bottom*bottom) }
  _R0 = (unsigned int)*(_BYTE *)(a1 + 2) >> 4;
  _R4 = 1000;
  __asm { SMLABB.W R0, R4, R0, R3 ; Signed multiply-accumulate (bottom*bottom) }
  _R2 = v2 & 0xF;
  _R3 = 100;
  __asm { SMLABB.W R2, R3, R2, R0 ; Signed multiply-accumulate (bottom*bottom) }
  return (v1 & 0xF) + _R2 + 10 * (v1 >> 4);
```

**Comment by rogerclarkmelbourne** [comment]
Ah. No worries. Leave it. Its pretty easy to do aoi(), I already wrote it, and its working. It just seemed like unnecessary duplication

**Comment by forkoz** [comment]
If it turns up later we can switch it out. And I'm guessing you'll be able to call the same AtoI in future mods.

**Comment by rogerclarkmelbourne** [comment]
Re: TG getting reset

It looks like 0x18bba which is the function used to setup various aspects of the TG,  i.e it has a switch statement and can set the number part or the TimeSlot , and some other non TG stuff...

Gets called by 0x18F76 a few seconds after the PTT is released

0x18F76 gets called indirectly via another function as a pointer to a function at address 0x13306 , in function 0x132B6

I put a breakpoint at 0x13306 and it gets called all the time, I suspect its part of the scheduler or on the timer tick callback.
The function pointers that 0x13306 uses are at 0x1fff5b98 and I set its set to 0x1e4b5 when somehow the code at 0x18F76 gets run.

**Comment by forkoz** [comment]
You mean 1BBA4?, 18bba is in the middle of a function. 0x13306 is in the middle too.

**Comment by rogerclarkmelbourne** [comment]
Yep
Sorry. I mean 

undefined FUN_0001bba4()

**Comment by forkoz** [comment]
Btw, in the ram dump I have  0x1fff5b98 -> sub_1E4B4, its a big table of pointers that starts at 1FFF5B8C

**Comment by rogerclarkmelbourne** [comment]
Thanks

**Comment by rogerclarkmelbourne** [comment]
BTW

tracing back up to what calls the channel change I get


```
channel_change_00012608
channel_change_00014252
channel_change_0001480e
channel_change_00014ebc
channel_change_00018f76
channel_change_0001eaa0
```

**Comment by rogerclarkmelbourne** [comment]
FUN_00012338 looks like its called when any of the non-keypad buttons is pressed

i.e It gets called if I press the PTT or the orange button, or the black or blue buttons, but the values in the registers is totally different depending on the button

**Comment by rogerclarkmelbourne** [comment]
FYI

Tx_end_00018cb0

**Comment by rogerclarkmelbourne** [comment]
I think I know approximately whats happening, there are a series of functions which each setup a callback (probably on a timer), which then call the next function in the chain

The TG gets reset by the last call to Channel_update_params_00018f76
This seems to get called from Tx_end_00018304(), 
```
        0001834a bd e8 10 40     pop.w      { r4, lr }
        0001834e 00 f0 12 be     b.w        Channel_update_params_00018f76                   undefined Channel_update_params_
                             -- Flow Override: CALL_RETURN (CALL_TERMINATOR)

```

But since the calling function Tx_end_00018304 pops  lr off the stack, when I look at the stack trace in 
Channel_update_params_00018f76
It doesn't show Tx_end_00018304  it shows the address in callback_execute_000132b6 from where Tx_end_00018304  was called  i.e

```
        00013304 e8 68           ldr        r0,[r5,#0xc]=>PTR_FUN_0001e4b4+1_1fff5b98        = 0001e4b5
        00013306 80 47           blx        r0
```

What sets up the initial call to Tx_end_00018304   is this code
```

                             LAB_00019062                                    XREF[1]:     0001905c(j)  
        00019062 60 7b           ldrb       r0,[r4,#0xd]=>DAT_1fff005d                       = 06h
        00019064 ff 28           cmp        r0,#0xff
        00019066 04 d1           bne        LAB_00019072
        00019068 df f8 e8 0b     ldr.w      r0=>Tx_end_00018304+1,[PTR_Tx_end_00018304+1_0   = 00018305
        0001906c fa f7 86 f8     bl         Callback_set_0001317c                            undefined Callback_set_0001317c()
```

which is in Channel_update_params_00018f76 its self
```

                             LAB_00019062                                    XREF[1]:     0001905c(j)  
        00019062 60 7b           ldrb       r0,[r4,#0xd]=>DAT_1fff005d                       = 06h
        00019064 ff 28           cmp        r0,#0xff
        00019066 04 d1           bne        LAB_00019072
        00019068 df f8 e8 0b     ldr.w      r0=>Tx_end_00018304+1,[PTR_Tx_end_00018304+1_0   = 00018305
        0001906c fa f7 86 f8     bl         Callback_set_0001317c                            undefined Callback_set_0001317c()
```

So I think what I need to do is possibly to modify Tx_end_00018304, so that it doesn't call    

`      0001834e 00 f0 12 be     b.w        Channel_update_params_00018f76                   undefined Channel_update_params_`


Unless the radio is in private call mode.

I'm not 100% sure this will work, because I've partially miss named Channel_update_params_00018f76 as it doesn't just change the channel TG parms, it also seems to act as a state machine.

But the only way to know is to patch the code and give it a try.

**Comment by forkoz** [comment]
FunctionButtons_1226C is called by 1261E which is called by the main task.  Its some sort of CB that isn't used if side buttons are held. The main task does the CPU dword function to see if the system is fully booted first, then runs this.

This stuff has to do with the 1,3,4,5,6 key values we get from the startup keys function. sub_171FA is checked and then we get the calibration backup/flash mode or startup_key_value = X (maybe should be function key?) 

I can speculate that this does the long press of the buttons when the system is up and running, but I've got no idea what all the memories are. EG word_1FFF00FC



```
int sub_1261E()
{
  int result; // r0@1

  sub_1F8CC();
  result = return_startup_key_value();
  if ( result )
  {
    if ( result == 5 || result == 6 )
    {
      sub_12772();
      if ( byte_1FFF19DB == 6 )
        byte_1FFF19DB = sub_20202((int)sub_124FA);
      result = sub_1E668();
      if ( result == -5 )
        result = sub_2041C();
    }
  }
  else
  {
    sub_12772();
    if ( byte_1FFF19DB == 6 )
      byte_1FFF19DB = sub_20202((int)FunctionButtons_1226C);
    result = sub_1E668();
    if ( result == 0xFFFFFFFB )
      result = sub_2041C();
  }
  return result;
```

**Comment by rogerclarkmelbourne** [comment]
I am going to have another look at Channel_update_params_00018f76

I think potentially there may be one section of the code, which modifies the TG, which is only run during the Tx End sequence.
In which case I can move that block into a separate function and check whether the radio is in the Manual Dial menu.

There is another block of code, which I think changes the TG flags, which I may not need to modify except if I provide a way of directly entering both the TG and the TS

**Comment by rogerclarkmelbourne** [comment]
I think I have a work-around / partial fix for the direct TG problem

This code is called both when you channel channel using the buttons, but also at the end of a transmission to set the TG back to the codeplug value.


```
                             LAB_00018f94                                    XREF[1]:     00018f8e(j)  
        00018f94 69 46           mov        r1,sp
        00018f96 17 20           mov        r0,#0x17
        00018f98 fb f7 20 fc     bl         FUN_000147dc                                     undefined FUN_000147dc()
        00018f9c 01 a9           add        r1,sp,#0x4
        00018f9e bd f9 00 00     ldrsh.w    r0,[sp,#0x0]=>local_20
        00018fa2 fc f7 25 f8     bl         FUN_00014ff0                                     undefined FUN_00014ff0()
        00018fa6 00 23           mov        r3,#0x0
        00018fa8 01 aa           add        r2,sp,#0x4
        00018faa 9d f8 08 10     ldrb.w     r1,[sp,#local_18]
        00018fae 0e 20           mov        r0,#0xe
        00018fb0 02 f0 f8 fd     bl         channel_set_params_0001bba4                      undefined channel_set_params_000
        00018fb4 e0 8a           ldrh       r0,[r4,#0x16]=>WORD_1fff0066                     = C0C0h
        00018fb6 40 04           lsl        r0,r0,#0x11
        00018fb8 40 0c           lsr        r0,r0,#0x11
        00018fba e0 82           strh       r0,[r4,#0x16]=>WORD_1fff0066                     = C0C0h
        00018fbc fd f7 9a ff     bl         FUN_00016ef4                                     undefined FUN_00016ef4()
        00018fc0 60 7b           ldrb       r0,[r4,#0xd]=>BYTE_1fff005d                      = 6h
        00018fc2 ff 28           cmp        r0,#0xff
        00018fc4 0e d0           beq        LAB_00018fe4
        00018fc6 02 23           mov        r3,#0x2
        00018fc8 00 22           mov        r2,#0x0
        00018fca 11 46           mov        r1,r2
        00018fcc 0a 20           mov        r0,#0xa
        00018fce 02 f0 e9 fd     bl         channel_set_params_0001bba4                      undefined channel_set_params_000
        00018fd2 00 22           mov        r2,#0x0
        00018fd4 0c 21           mov        r1,#0xc
        00018fd6 02 20           mov        r0,#0x2
        00018fd8 f9 f7 93 fc     bl         keypad_00012902                                  undefined keypad_00012902()
        00018fdc 04 f1 0d 00     add.w      r0=>BYTE_1fff005d,r4,#0xd                        = 6h
        00018fe0 fa f7 02 f9     bl         Tx_end_delayed_000131e8                          undefined Tx_end_delayed_000131e
```

Specifically 

 00018fb0 02 f0 f8 fd     bl         channel_set_params_0001bba4                      undefined channel_set_params_000

So I commented that line out.

And surprisingly the radio still worked fine for normal changing of channels from the menu keys etc

So its function must be duplicated elsewhere

Strangely, everything seems to work just fine !

But I think the best thing to do is not just to comment this line out with NOPs, but probably to call a wrapper function which does some additional checks

because the lower half of that block of code does this check

```
  00018fc0 60 7b           ldrb       r0,[r4,#0xd]=>BYTE_1fff005d                      = 6h
        00018fc2 ff 28           cmp        r0,#0xff
        00018fc4 0e d0           beq        LAB_00018fe4
```

Which seems to be a flag to say its being run because its in the Tx_end sequence

So what I can do in the wapper function is basically use that code and also check if the radio is in manual dial mode.

ie. 
```
if (!Tx_END || (Tx_END && Manual_Dial_Mode)
{
channel_set_params_0001bba4 ()
}
```

But I'm tempted just to comment the call out, and let some people try it, and see if the call to channel_set_params_0001bba4  is redundant

**Comment by forkoz** [comment]
So direct TG entry will?
1. Get rid of private call mode, eh *we* don't use it but commercial users might. NBD
2. Replace the TXTG in the codeplug?
3. Replace the TXTG in the conact?
4. Values stay through poweroff?

**Comment by rogerclarkmelbourne** [comment]
1.Brandmeister and other networks support private call, so it needs to be retained
2. I could do this.  But people probably won't want to permanently mess with their codeplug, unless we changed the whole structure and did away with the need to duplicate the channel multiple times, each with a different TG
3 It doesn't update a Contact
4. Not not unless I write it into the codeplug (see item 2 ;-)

**Comment by rogerclarkmelbourne** [comment]
Its really odd at the moment.

With this line commented out, I think its sending an All Call.

I have 2 GD-77s and when I transmit, it doesnt matter what I type into the Manual Dial, my other radio always says its being called

**Comment by forkoz** [comment]
So only stays till you switch the channel or till you shut down?

**Comment by rogerclarkmelbourne** [comment]
Yes

**Comment by forkoz** [comment]
Both? That's not bad. Mess with the TG, which duplicates to RX and talk and then restores when you move on.  I almost want it to stay till power off (even walking away from the channel) but I think that might cause problems anyway.

**Comment by rogerclarkmelbourne** [comment]
Actually. With that line commented out. Once you make a private call, you are stuck in PC mode until you reboot the radio.

So I do need to make a patch function to wrap the call and make it when its in Manual Dial mode

Other times it doesn't seem to need the call to be made

**Comment by rogerclarkmelbourne** [comment]
argghh

Sometimes it seems to be saving the name back into the codeplug when I turn it off and on again

But its not saving the TG into the codeplug.

**Comment by rogerclarkmelbourne** [comment]
I just read it back into the CPS and its updated the name to TG555 but not changed anything else

![image](https://user-images.githubusercontent.com/6146984/55921712-4ca4a900-5c41-11e9-97ac-e28d81ace336.png)

**Comment by forkoz** [comment]
On mine it jumps to the 1st channel in the zone list on every power off.

**Comment by rogerclarkmelbourne** [comment]
Yep. Sometimes mine starts up in VFO mode, even though I left it in Channel mode.

BTW. I think I can guess why it can't save the TG. I bet if I manually set a TG which is in the Contact list it would change it, as I've notice it seems to look up the Contact from the TG Number when you transmit

So I'll try changing to a TG thats in a contact, and see if the TG then gets saved.

I really don't see any need for the channel name etc to be resaved each time the radio is powered down.

Again, its really crappy coding

**Comment by rogerclarkmelbourne** [comment]
Na

Thats didnt work ;-)

**Comment by forkoz** [comment]
Playing around I got the radio to start scanning and not stop after multiple reboots, like scan was running but it said scan was off. Top line was just scrolling through screen until I rewrote the code plug. So there may be more bugs yey. Definitely going to have a backup CP beta testing this too.

We do have to see what the heck the shutdown task is doing... perhaps its not delaying long enough to save all parameters?

**Comment by rogerclarkmelbourne** [comment]
Yep.
The powerdown bug was on my list, but its now gained priority ;-)

**Comment by rogerclarkmelbourne** [comment]
[direct_TG_entry_partial_patches.zip](https://github.com/rogerclarkmelbourne/Radioddity_GD-77/files/3066366/direct_TG_entry_partial_patches.zip)

These 2 files patch in the direct TG entry. i.e they catch the # key when in direct dial mode and then transfer that TG to the screen as TGxx and also to the actual TG memory number.

However, to get it to kinda work, you also need to manually nop out

`        00018fb0 02 f0 f8 fd     bl         channel_set_params_0001bba4                      undefined channel_set_params_000`

because that line sets the TG back to the codeplug value.

Note. After patching that line, you can't exist from PC unless you reboot.

I need to do a wrapper function for that line, but I don't have time to do it now

**Comment by forkoz** [comment]
No prob. I'm probably going to wait on the fix for code plug sorcery and see what's up with the shutdown handler.

**Comment by rogerclarkmelbourne** [comment]
OK

**Comment by rogerclarkmelbourne** [comment]
BTW.
I put a read watch on 0x1fff5940 where I wrote the TGxxx for display on the screen.

And the read watch got triggered by a memcpy, which was called from here

`        0001e5bc 00 f0 80 fe     bl         memcpy_0001f2c0                                  void memcpy_0001f2c0(int iParm1,`


I'm not sure if the debugger is telling the truth about the memory being read.

But if i turn off the radio, I don't get any read accesses. So I suspect its reading the data when you change channel and using that data sometimes when the radio powers down

**Comment by forkoz** [comment]
Dword right obove there: dword_1FFF5938 is used by  Shutdown_storing_1E48A

It does like this:

```
int *Shutdown_storing_1E48A()
{
  signed int v0; // r4@1
  int *result; // r0@2
  int v2; // r0@3
  int v3; // r2@3
  int *v4; // r3@3

  for ( LOBYTE(v0) = 0; ; LOBYTE(v0) = v0 + 1 )
  {
    v0 = (unsigned __int8)v0;
    if ( (signed int)(unsigned __int8)v0 >= 8 )
      break;
    result = dword_1FFF5938;
    if ( *((_BYTE *)dword_1FFF5938 + (unsigned __int8)v0) == 0xAA )
    {
      *((_BYTE *)dword_1FFF5938 + (unsigned __int8)v0) = 0;
      v2 = sub_1E290(v0);
      result = write_out_eeprom_1E394(v2, v0, v3, v4);
    }
  }
  return result;
}
```

I wonder if there is simply not enough delay to write everything out:

```
int *Task_0x03_shutdown()
{
  int v0; // r4@1
  __int64 v1; // r0@3
  __int64 v2; // r2@3

  v0 = 0;
  do
  {
    lwsem_wait((int)&ShutdownSem_1FFF135C);
    do
    {
      if ( !sub_13366(0) )
        break;
      v1 = time_delay_ticks(0xAu, SHIDWORD(v1), v2);
      LOWORD(v1) = v0++;
    }
    while ( (signed int)(unsigned __int16)v1 < 300 );
    time_delay_ticks(1u, SHIDWORD(v1), v2);
  }
  while ( GreaterThan_24126(0, 0xD) != 1 );
  return Shutdown_storing_1E48A();
}
```

**Comment by rogerclarkmelbourne** [comment]
One way to test this is to reduce the delay even further and see if the problem is worse.

BTW. Can we move this to a separate issue.

I know its related, but its also a bug in its own right

**Comment by rogerclarkmelbourne** [comment]
I'll make an issue

**Comment by rogerclarkmelbourne** [comment]
https://github.com/rogerclarkmelbourne/Radioddity_GD-77/issues/39

**Comment by rogerclarkmelbourne** [comment]
BTW. What address is Task_0x03_shutdown

**Comment by rogerclarkmelbourne** [comment]
I've written the wrapper patch to prevent the direct entry TG getting over written.

Here are the 2 files. 
One simply replaces the call to the original function with a call to the wrapper function, which calls the original function if needs to do so (but not when it would overwrite the TG)

[Patches_to_stop_direct_entry_tg_getting_overwritten.zip](https://github.com/rogerclarkmelbourne/Radioddity_GD-77/files/3068169/Patches_to_stop_direct_entry_tg_getting_overwritten.zip)

**Comment by rogerclarkmelbourne** [comment]
However, I still have the problem where sometimes the channel name from the display gets written back into the codeplug.

Its really random when it does this.

But whats really concerning is that if I put a breakpoint on the 0x26072, which seems to be the SPI Flash write function, it calls it alarming often, and I because I have been resetting in the debugger, I send up with pages in flash which have been erased but not rewritten

The reason its alarming is because Flash memory has a finite number of erase cycles. The data sheet lists this as "More than 100,000 erase / program cycles" so I guess its not terrible that it seems to do this every time the radio is turned on. 

But it does seem to do this unnecessarily, which will be wearing out the flash when it doesn't need to.

**Comment by forkoz** [comment]
Does it still call it so many times on stock firmware?

**Comment by rogerclarkmelbourne** [comment]
I don't know.

I'll revert and check.

It looks like something is setting up a callback so that Shutdown_storing_1E48A gets called at seemly random times.

Also Shutdown_storing_1E48A seems to be called at least twice during startup. But I don't think it ends up writing to flash, as I put a breakpoint in 0x26072 which is the business end of the Flash write and it doesnt get triggered at startup

It does seem to cause more flash writes if I go into manual dial. 
Even just starting to type a number causes it to write to flash.
This is before its even using my new code

**Comment by rogerclarkmelbourne** [comment]
I just tried the official firmware and it seems to be roughly the same. Just entering the private call area seems to trigger it after a while

**Comment by forkoz** [comment]
the private call stuff triggering any flash (or ee) writes in the code?

**Comment by rogerclarkmelbourne** [comment]
Seems to cause some sort of backup.

0x26072 seems to be called multiple times with blocks of data. i.e I think it may rewrite most of the codeplug flash.
But its hard to debug because if I leave it too long, I think the watchdog or something else kicks in and I end up with big blocks of erased flash

I think its a known bug. Well, I didnt know what caused it. But the radio displays something like "No channel" when the flash has been erased.

And I know other people have reported this.

I think we need to overhaul the flash Writes, as I can't see its necessary to do it all the time.

It really only needs to write stuff at shutdown.

And we'd always presumed it was using EEPROM to store the state of the radio, but it looks like it could be using flash as well :-(

BTW. Its getting late here, so I'm calling it a day now. ;-)

**Comment by rogerclarkmelbourne** [comment]
Tomorrow morning I'll try nop'ing out whatever function is causing the flash writes to happen, and then see what the side effects are.

I presume it simply won't save the last channel is was on, but it doesn't do that correctly even with writing the flash a lot ;-)

**Comment by rogerclarkmelbourne** [comment]
I am going to put a Write Watch on the channel names in the codeplug in memory and seen when the radio updates them.

In put a Read Watch on the active channel name memory but it gets read loads of time but the function that reads the memory does a lot of other functions, and seems to write the channel name string elsewhere.

So I think Breaking when the RAM copy of current zone’s Channels is written, is probably going to be the most productive angle

**Comment by rogerclarkmelbourne** [comment]
My attempt to prevent the TG getting restored when you finish transmitting, apart from when in Manual Dial mode doesn't seem to always work.

I'm going to give up on modifying their hack, and do it a different way.

Rather than the code always re-instating the TG and the TS / mode flag every time the Ptt is released, it would be better to put the code that re-instates the codeplug value, into the code that handles leaving the Manual Dial system

**Comment by rogerclarkmelbourne** [comment]
FYI. I'm having to rethink the way to do the direct TG entry as I'm constantly fighting against the functionality where it keeps trying to set the TG back to the TG assigned to the Channel

I've now found this function

Codeplug_get_TG_for_channel_0001b228

Which seems to get passed something like the Channel index in the Zone and some other information, and returns the TG number for that channel.

I'm going to patch everywhere that uses this function to use a wrapper function which first checks whether an override TG has been entered and if so returns that TG, and otherwise it returns the TG for the channel.
The only think I need to make sure of is that I clear the override TG when the user either presses the Channel up or down buttons or if they change zones.

The Channel up and down buttons are easy to patch, but the Zone selection is more difficult, as I'll need to find the Menu code that handles the Zone changing.

**Comment by rogerclarkmelbourne** [comment]
Believe it or not, i am still working on this, and have put at least 10 hours into it this week alone.

It has proved to be much more difficult than I imagined.

The main problem is the TG or PC value that keeps getting put into the master TG/PC/TS control variable ( 32 bit memory location)

I can easily update this variable and set a new TG, just by changing the lower 24 bits.
I leave the top 8 bits alone, since it contains the TimeSlot and also flags about whether it’s an Private Call or a Grouo Call ( or I suppose an All Call)

What I had not realised is that this variable is changes a lot by other parts of the firmware.

The Private Call system updates this variable when you press the PTT, wirh the target DMR ID as well flags to indicate it’s a PC.
I think there is a flag variable elsewhere which may also indicate whether it’s a PC or. Group Call but I have not tracked it down, ( and it may not exist at all)

Normally after the PTT is released the firmware reinstates the original channel’s TG and TS back into the master variable, hence my code also needs to handle putting the overrride TG back into the master variable 

But also, what I had not realised is that the master TG / PC variable also gets updated when there is an incoming signal which has been accepted, I.e is in the RxGroup list or is the users own DMR ID, for a PC.

This is because of the strange feature in DMR where if you receive a Call ( either TG or PCj and reply within a soecificed period of time, the radio automatically changes TG to the incoming TG or PC number.

For Ham Radio use, this “feature” of changing TG without you knowing it, can be problematic and is probably the reason people are told not to use Rx Groups, ( or just 1 TG per Rx Group) as if you can then only receive one TG hence can’t unexpected transmit in another TG.

I can now easily stop this “feature” happening by pactching the functionality which changes the TG when there is an incoming call.
However think I will leave this alone at the moment, because it would probably still need to apply to incoming PCs

I found a hacky way to prevent the TG getting updated with the original Channel’s TG by accident, because I didn’t realise I was overriding the function which converted the BCD representation of the TG to an int 
This worked pretty well, but since the users ID and other stuff seems to be stored in BCD, it seems like a sledgehammer to crack a nut and may have unintended consequences , if I simply always returned the overridden TG ( is there was one) from the BCD2Int function.

Currently to make this hack work, I store the old TG that was assigned to the channel and the override TG, and I compare what the BCD2Int function returns and if it’s the old TG, and I have an override TG, I return the override TG instead.

I think I could actually put this hack elsewhere and simply patch the one bit of code which I think the firmware always uses to set the master TG variable.

But again, it’s a hacky solution and overkill.

So I am now looking at other options.
I tried to find where in the memory it stores the data for the current channel from the codeplug.

But I found out that it doesn’t store it. It keeps reading it from the Flash (or EEPROM) if it needs the TG for the channel.

This was a bit surprising, since I assumed they’d load the current channel TG and Freq etc as a binary struct so that it could be easily accessed, but AFIK they don’t do that.

So I could override the function to get current channel data and return a differnt TG, but again this seems like overkill.

So, am now investigating what I thiink is the one area of the code which puts the channel TG back into the master TG variable 

But at this point, the TG is in BCD format as it’s normally just been retrieved from the codeplug.

So what I think I need to do is to convert the Override TG into BCD and inject it at this point.

BTW. The reason we can’t seem to find an atoi() function, is because they never seem to need one.

I think there must be an aToBCD() function somewhere but I think it will be quicker for me to write the function than to spend ages looking it.

I am hoping finally this may be the optimal solution for the Manual TG entry system

I will let you know if this works.

**Comment by forkoz** [comment]
Wow. I 've been using multiple TGs in each RX group and it has always either replied to transmissions and after the call has ended gone back to the TX tg set for the channel. I can also pull up a contact and then transmit on that when it is selected in "contacts".

**Comment by rogerclarkmelbourne** [comment]
I'm not going to change any of that functionality.

Hence things are a bit complicated because I need to only intercept and change master TG update, if its attemping to to re-instate the TG for the current channel when I have overridden the TG for that channel

One of the places I need to do this is

`        00018fb0 02 f0 f8 fd     bl         Set_Channel_params_0001bba4                      undefined Set_Channel_params_000
`
because it has this just before

`        00018fa2 fc f7 25 f8     bl         Coceplug_get_TG_BCD_for_channel-014ff0           undefined Coceplug_get_TG_BCD_fo`

But there may be other places like this in the code

I'm going to need to write a wrapper function for Set_Channel_params_0001bba4    which checks whether i have overridden the TG and if so use the override BCD value instead of the one from the codeplug

**Comment by forkoz** [comment]
Definitely hope you post function names dump when you're done.  I still have a blank on most of the non-os stuff. BTW, found a ptr table similar to the menu that points at the icons/graphics @ 0x413D0. Unlike the menu the graphics have a visual guide at least :)

**Comment by rogerclarkmelbourne** [comment]
Yes.. I found that Ptr table as well, and have updated on of my Ghidra saved revisions. 

But I managed to screw up something in my later changes in Ghidra and had to revert back to an older version, so the version that contained the menu stuff is currently orphaned.

I'll need to somehow export the menu function names from the orphaned version into the version I'm now using, but in the orphaned version where I'd renamed the menu functions, I also had to use the Create function to turn some of the labels into functions, hence its all a bit complicated to merge it :-(

**Comment by forkoz** [comment]
Send over the db if you can. I can compare with the table and probably track that stuff down.

**Comment by rogerclarkmelbourne** [comment]
I can put a ghidra backup somewhere but it need to be a private transfer. 
If you can remember what email you put into the comment on my blog, I can setup an FTP account with forkoz as your username and that email add as the password.  I assume it was not a valid email add ?

**Comment by forkoz** [comment]
It was a temp email, I lost it :( you can send to nsuvrvmx@sharklasers.com for at least 20-30 if that is enough time

**Comment by rogerclarkmelbourne** [comment]
OK.

I'll export a ghidra backup, and emial it to that add

It won't have a working direct entry of TG, but it will have loads of menu functions defined, 

Well, at least that they are a menu function  of some kind ;-)

**Comment by rogerclarkmelbourne** [comment]
OK. Sending it now. Its 3Mb

Just import it into Ghidra and look under m for menu.
There are stacks of functions, and you can cross reference them back to the menu pointer table

Its a real pain that I screwed something else up in this version, and had to go back 1 week :-(

**Comment by rogerclarkmelbourne** [comment]
I'm now changing my workflow

Rather than making small patches by exporting part of Ghidra, its safer to export the whole of ghidra as intel hex, then I don't get out of sync between the various patches,

I'm not even absolutely sure the Rx Group stuff is in the version I emailed.
It should be, but I'm getting overwhelmed with the number of small changes need in the code.

**Comment by rogerclarkmelbourne** [comment]
arrrg

That email add has a max attachment size of 1Mb :-(  It was 4mb

**Comment by forkoz** [comment]
send ftp and password there :)

**Comment by rogerclarkmelbourne** [comment]
Good idea

I'll setup an account now.

**Comment by forkoz** [comment]
What I've been doing is making a backup of the file before patching. In Ida I can reload the renamed input file and then bad patches disapear. Also because ida is shit at undoing things previously patched, they might not actually get written out.  Plus I work on one copy stock and then 1 copy patched to ease seeing what was there before.

**Comment by forkoz** [comment]
I just ftp@ your domain?

**Comment by rogerclarkmelbourne** [comment]
I have been backing up Ghidra every day, but I got out of sync.

Ideally everthing should be in git with some sort of diff mechanism.

**Comment by rogerclarkmelbourne** [comment]
Check your email now.

**Comment by forkoz** [comment]
yep, got it, filezilla is being a pain

**Comment by forkoz** [comment]
How do I open GZF?

**Comment by rogerclarkmelbourne** [comment]
Its a ghidra backup

Import into ghidra

**Comment by forkoz** [comment]
Thanks, I'll go through it today/tomorrow.

**Comment by rogerclarkmelbourne** [comment]
No worries. I decided to go through my latest version, and just convert the functions in the menu table which Ghidra had miss-categorized as Labels,

Hopefully I should be able to export the menu functions symbols list from the version you downloaded, and import that into my latest version, so at least I will have all the menu functions named as menu-xxxxxx

I may actually rename them to Menu-Item-xxxx as I think since its a pointer table the function pointers probably relate to individual menu options.

But I have yet to work out how the pointers map to the menu numbers.

BTW. I've got the keypad entry of TG working quite well, but I need to clear the override if the user selects a new zone and I don't know where the zone selection code is handled yet

**Comment by forkoz** [comment]
Almost ready for prime time. I also see the variable lenght DMRID patch in there.

**Comment by rogerclarkmelbourne** [comment]
BTW. I think in that version I changed the memory access keys, so you only need to hold Blue and the * key, as I got fed up with needing to press 3 keys, but think I'll change it back as will be confusing for people to have 2 different key sequences.

Ultimately there is no need for the key sequence at all

**Comment by forkoz** [comment]
I did a similar patch on the 1801, just SK2 and 0, it also kills those calibration backups by never going to them.

**Comment by rogerclarkmelbourne** [comment]
I’ve realised that getting direct TG entry to work in the Doubke Wait modes would take ages, so I may add a check to prevent people being able to enter that mode.

However, even finding out where the mode information is stored is proving difficult 

I even wrote a simple C# application to compare multiple memory dumps to look for similarities and differences , but there are still multiple places in memory which differ only because of the Double Wait system.

Since it doesn’t do any harm in double wait mode, I may have to give up disabling it and people will just need to not try to use it at the moment.

However I still have 2 other things I need to work out for the direct TG stuff

1. Why the memory that contains the string “VFO” doesn’t alwsys get updated with the TG xxxx text
2. I need to re-enable channel save on shutdown, but I need to fix it first (aarrggghhh) as it doesn’t alwsys save and also saves stuff it should not.

**Comment by forkoz** [comment]
I use double single, so I have 2 vfos but not double double. Will it not work?

**Comment by rogerclarkmelbourne** [comment]
I've not tried Single Wait

What seems to happen in Double Wait is that the original TG is restored after the transmission has finished.

There must be another path though the code that its using to restore the TG.

I only patched the TG restore function call to go to my wrapper function in one place.

I guess I'll need to find the other places it does this.

**Comment by forkoz** [comment]
Try in single wait, it may just work.... I hope

**Comment by rogerclarkmelbourne** [comment]
Looks like it partially works.

If you set the TG of the channel at the top of display (lets call it channel A)
Then it all works fine.

If you change the the TG of channel B, it overrides it, but it updates the display for channel A not channel B.

I presume in all the Double wait modes they have 2 display buffers one for channel A and the other for channel B.

If you change the active channel from A to B, it uses the TG for channel B, and when you switch back to channel A, that channel goes changes its display back to the name and the TG for the channel.
i.e it looks like it probably loads the data for both channel A and B when you switch between A and B

It would be possible  to get it to work fully, but it would take a load more work, and I think there are more important things i need to look at.

**Comment by forkoz** [comment]
That seems like it mostly works. If you switch back/forth the changes are cleared just like they would be if you changed channel or zone, you effectively did. What's it look like when it updates the buffer of channel A when you're on B? Top line changes to the tg? Or you will press PTT on B and have the wrong TG displayed?

**Comment by rogerclarkmelbourne** [comment]
When you change from A to B it updates the Master TG location to the TG for channel B.

It doesnt update the display for channel A, so that stays at the value I assigned e.g. TG 999
But when you switch back to channel A, it updates both the display with the TG for the channel and also updates the TG in the master location.

I presume in Single Wait, that only one channel is active at a time, so as long as you only use direct TG entry in channel A it would appear to work OK, except it will change TG back to the original TG when you change between A and B.

I'm sure it would be possible to get it to fully work, but it would probably require quite a major overhaul of the way the TG and channel info is handled.
i.e I am suprised that they don't cache the current channel information into RAM. They retrieve it from the codeplug every time that they need to use it.

Hence I can't simply change a cached version.

One possible solution is to store the zone and channel that have been overridden and then intercept the codeplug read and check if the read is for the overridden channel.
But when I looked at that method, it wasn't entirely clear how the firmware retrieves the channel data from the codeplug.

It looks like its done in 

Coceplug_get_TG_BCD_for_channel-014ff0

as this is called just prior to where it makes the call to restore the TG back to the value for the channel
```

        00018f9c 01 a9           add        r1,sp,#0x4
        00018f9e bd f9 00 00     ldrsh.w    r0,[sp,#0x0]=>local_20
        00018fa2 fc f7 25 f8     bl         Coceplug_get_TG_BCD_for_channel-014ff0           undefined Coceplug_get_TG_BCD_fo
        00018fa6 00 23           mov        r3,#0x0
        00018fa8 01 aa           add        r2,sp,#0x4
        00018faa 9d f8 08 10     ldrb.w     r1,[sp,#local_18]
        00018fae 0e 20           mov        r0,#0xe
        00018fb0 4c f0 a2 fa     bl         Patch_Set_Channel_params_000654f8                undefined Patch_Set_Channel_para

```

**Comment by rogerclarkmelbourne** [comment]
I found one other place that the TG is set 

at 0x19030  but I don't think I'll patch that call at the moment, because I'm not sure what the code is doing at that point.

I want to patch as little as possible with this, as it could otherwise have unintended consequences

**Comment by forkoz** [comment]
Yep, learning that the hard way with the IME stuff and extended DMR ids. Eventually all the unintended stuff will come out.

**Comment by rogerclarkmelbourne** [comment]
I'll need to harden  the existing DMR ID code.

**Comment by forkoz** [comment]
Another bug I saw recently was FFF as the # below the name when private calling. Long DMR ID text was fine and the call makes it.

**Comment by rogerclarkmelbourne** [comment]
I think it’s a bug in the official firmware
